File: backend/app/services/notification_rules.py
from datetime import date, timedelta

from dateutil.relativedelta import relativedelta
from sqlalchemy import func
from sqlalchemy.orm import Session

from app import models
from app.config import get_settings

settings = get_settings()

MILESTONES = (25, 50, 75, 100)


def _create_if_new(db: Session, household_id: str, dedupe_key: str, **kwargs) -> models.Notification | None:
    existing = (
        db.query(models.Notification)
        .filter(models.Notification.household_id == household_id, models.Notification.dedupe_key == dedupe_key)
        .first()
    )
    if existing:
        return None
    notification = models.Notification(household_id=household_id, dedupe_key=dedupe_key, **kwargs)
    db.add(notification)
    return notification


def check_budget_overages(db: Session, household_id: str) -> list[models.Notification]:
    created = []
    month_start = date.today().replace(day=1)
    month_end = month_start + relativedelta(months=1)
    period_label = month_start.strftime("%Y-%m")

    budgets = db.query(models.Budget).filter(models.Budget.household_id == household_id, models.Budget.month == month_start).all()
    for budget in budgets:
        spent = (
            db.query(func.coalesce(func.sum(models.Transaction.amount), 0))
            .join(models.Account)
            .filter(
                models.Account.household_id == household_id,
                models.Transaction.category_id == budget.category_id,
                models.Transaction.direction == models.TransactionDirection.debit,
                models.Transaction.date >= month_start,
                models.Transaction.date < month_end,
            )
            .scalar()
        )
        spent = float(spent or 0)
        limit = float(budget.amount_limit)
        if limit <= 0:
            continue
        pct = spent / limit
        category_name = budget.category.name if budget.category else "this category"

        if pct >= 1.0:
            note = _create_if_new(
                db,
                household_id,
                dedupe_key=f"budget:{budget.id}:{period_label}:exceeded",
                type=models.NotificationType.budget_exceeded,
                severity=models.NotificationSeverity.warning,
                title=f"Budget exceeded: {category_name}",
                message=f"You've spent ${spent:,.2f} of your ${limit:,.2f} {category_name} budget this month ({pct * 100:.0f}%).",
                related_entity_type="budget",
                related_entity_id=budget.id,
            )
            if note:
                created.append(note)
        elif pct >= settings.budget_warning_pct:
            note = _create_if_new(
                db,
                household_id,
                dedupe_key=f"budget:{budget.id}:{period_label}:warning",
                type=models.NotificationType.budget_exceeded,
                severity=models.NotificationSeverity.info,
                title=f"Approaching budget limit: {category_name}",
                message=f"You've used {pct * 100:.0f}% of your ${limit:,.2f} {category_name} budget this month (${spent:,.2f} spent).",
                related_entity_type="budget",
                related_entity_id=budget.id,
            )
            if note:
                created.append(note)
    return created


def check_goal_milestones(db: Session, household_id: str) -> list[models.Notification]:
    created = []
    goals = db.query(models.Goal).filter(models.Goal.household_id == household_id).all()
    for goal in goals:
        if float(goal.target_amount) <= 0:
            continue
        pct = (float(goal.current_amount) / float(goal.target_amount)) * 100
        reached = [m for m in MILESTONES if pct >= m and m > goal.last_milestone_pct]
        if not reached:
            continue
        milestone = max(reached)
        severity = models.NotificationSeverity.success if milestone == 100 else models.NotificationSeverity.info
        title = f"Goal complete: {goal.name}!" if milestone == 100 else f"{milestone}% of the way to {goal.name}"
        note = _create_if_new(
            db,
            household_id,
            dedupe_key=f"goal:{goal.id}:{milestone}",
            type=models.NotificationType.goal_milestone,
            severity=severity,
            title=title,
            message=f"{goal.name} is at ${float(goal.current_amount):,.2f} of ${float(goal.target_amount):,.2f} ({pct:.0f}%).",
            related_entity_type="goal",
            related_entity_id=goal.id,
        )
        goal.last_milestone_pct = milestone
        if note:
            created.append(note)
    return created


def check_upcoming_bills(db: Session, household_id: str) -> list[models.Notification]:
    created = []
    today = date.today()
    lead_days = settings.upcoming_bill_lead_days
    bills = (
        db.query(models.RecurringBill)
        .filter(models.RecurringBill.household_id == household_id, models.RecurringBill.is_active.is_(True))
        .all()
    )
    for bill in bills:
        try:
            due_date = today.replace(day=bill.due_day)
        except ValueError:
            continue
        if due_date < today:
            due_date = due_date + relativedelta(months=1)
        days_until = (due_date - today).days
        if days_until < 0 or days_until > lead_days:
            continue

        period_start = due_date - relativedelta(months=1)
        already_paid = (
            db.query(models.Transaction)
            .join(models.Account)
            .filter(
                models.Account.household_id == household_id,
                models.Transaction.direction == models.TransactionDirection.debit,
                models.Transaction.date >= period_start,
                models.Transaction.date <= due_date + timedelta(days=1),
                models.Transaction.amount >= float(bill.amount) * 0.85,
                models.Transaction.amount <= float(bill.amount) * 1.15,
                *([models.Transaction.category_id == bill.category_id] if bill.category_id else []),
            )
            .first()
        )
        if already_paid:
            continue

        note = _create_if_new(
            db,
            household_id,
            dedupe_key=f"bill:{bill.id}:{due_date.isoformat()}",
            type=models.NotificationType.bill_due,
            severity=models.NotificationSeverity.warning if days_until <= 1 else models.NotificationSeverity.info,
            title=f"{bill.name} due {('today' if days_until == 0 else f'in {days_until} days')}",
            message=f"${float(bill.amount):,.2f} for {bill.name} is due on {due_date.strftime('%b %d')}.",
            related_entity_type="recurring_bill",
            related_entity_id=bill.id,
        )
        if note:
            created.append(note)
    return created


def check_large_transactions(db: Session, household_id: str) -> list[models.Notification]:
    created = []
    threshold = settings.large_transaction_threshold
    since = date.today() - timedelta(days=2)
    txns = (
        db.query(models.Transaction)
        .join(models.Account)
        .filter(
            models.Account.household_id == household_id,
            models.Transaction.direction == models.TransactionDirection.debit,
            models.Transaction.amount >= threshold,
            models.Transaction.date >= since,
        )
        .all()
    )
    for txn in txns:
        note = _create_if_new(
            db,
            household_id,
            dedupe_key=f"large_txn:{txn.id}",
            type=models.NotificationType.large_transaction,
            severity=models.NotificationSeverity.info,
            title=f"Large transaction: ${float(txn.amount):,.2f}",
            message=f"{txn.name} on {txn.date.strftime('%b %d')} was ${float(txn.amount):,.2f}.",
            related_entity_type="transaction",
            related_entity_id=txn.id,
        )
        if note:
            created.append(note)
    return created


def check_idle_cash_opportunities(db: Session, household_id: str) -> list[models.Notification]:
    created = []
    threshold = settings.idle_cash_investment_threshold
    period_label = date.today().strftime("%Y-%m")
    savings_accounts = (
        db.query(models.Account)
        .filter(
            models.Account.household_id == household_id,
            models.Account.type == models.AccountType.savings,
            models.Account.current_balance >= threshold,
        )
        .all()
    )
    for account in savings_accounts:
        note = _create_if_new(
            db,
            household_id,
            dedupe_key=f"idle_cash:{account.id}:{period_label}",
            type=models.NotificationType.investment_opportunity,
            severity=models.NotificationSeverity.info,
            title=f"Idle cash in {account.name}",
            message=(
                f"{account.name} has ${float(account.current_balance):,.2f}, above your "
                f"${threshold:,.2f} idle-cash threshold. Once your emergency fund is covered, "
                f"consider moving the excess into an investment or high-yield account."
            ),
            related_entity_type="account",
            related_entity_id=account.id,
        )
        if note:
            created.append(note)
    return created


def check_credit_alerts(db: Session, household_id: str) -> list[models.Notification]:
    created = []
    period_label = date.today().strftime("%Y-%m")
    profiles = (
        db.query(models.CreditProfile)
        .join(models.User)
        .filter(models.User.household_id == household_id)
        .all()
    )
    for profile in profiles:
        if profile.utilization_pct is not None and float(profile.utilization_pct) > 30:
            note = _create_if_new(
                db,
                household_id,
                dedupe_key=f"credit_util:{profile.user_id}:{period_label}",
                type=models.NotificationType.credit_alert,
                severity=models.NotificationSeverity.warning,
                title=f"High credit utilization for {profile.user.full_name}",
                message=(
                    f"{profile.user.full_name}'s reported utilization is {float(profile.utilization_pct):.0f}%, "
                    "above the 30% level that most scoring models penalize. Paying down balances "
                    "before the statement closes date can help."
                ),
                related_entity_type="credit_profile",
                related_entity_id=profile.id,
            )
            if note:
                created.append(note)

        last_two = (
            db.query(models.CreditScoreHistory)
            .filter(models.CreditScoreHistory.user_id == profile.user_id)
            .order_by(models.CreditScoreHistory.recorded_date.desc())
            .limit(2)
            .all()
        )
        if len(last_two) == 2 and (last_two[1].score - last_two[0].score) >= 15:
            note = _create_if_new(
                db,
                household_id,
                dedupe_key=f"credit_drop:{profile.user_id}:{last_two[0].recorded_date.isoformat()}",
                type=models.NotificationType.credit_alert,
                severity=models.NotificationSeverity.warning,
                title=f"Credit score drop for {profile.user.full_name}",
                message=(
                    f"{profile.user.full_name}'s score fell from {last_two[1].score} to {last_two[0].score}. "
                    "Check for new derogatory marks, hard inquiries, or a spike in utilization."
                ),
                related_entity_type="credit_profile",
                related_entity_id=profile.id,
            )
            if note:
                created.append(note)
    return created


def run_all_checks(db: Session, household_id: str) -> list[models.Notification]:
    created: list[models.Notification] = []
    created += check_budget_overages(db, household_id)
    created += check_goal_milestones(db, household_id)
    created += check_upcoming_bills(db, household_id)
    created += check_large_transactions(db, household_id)
    created += check_idle_cash_opportunities(db, household_id)
    created += check_credit_alerts(db, household_id)
    db.commit()
    for note in created:
        db.refresh(note)
    return created

File: functions/src/notificationRules.ts
// The notification rule engine: given a household's data and today's date, decides which
// notifications should exist. Pure (no Firestore), so it's easy to test; household.ts loads
// the data and writes the results.
import {
  BUDGET_WARNING_PCT,
  IDLE_CASH_INVESTMENT_THRESHOLD_CENTS,
  LARGE_TRANSACTION_THRESHOLD_CENTS,
  UPCOMING_BILL_LEAD_DAYS,
} from "./config.js";
import { addDays, addMonths, daysBetween, monthStart, shortDate, withDay } from "./dates.js";
import { money, wholePercent } from "./format.js";

export interface HouseholdData {
  accounts: { id: string; name: string; type: string; current_balance_cents: number }[];
  transactions: {
    id: string;
    account_id: string;
    category_id: string | null;
    date: string;
    amount_cents: number;
    direction: "debit" | "credit";
    name: string;
  }[];
  categories: { id: string; name: string }[];
  budgets: { id: string; category_id: string; month: string; amount_limit_cents: number }[];
  goals: { id: string; name: string; target_amount_cents: number; current_amount_cents: number; last_milestone_pct: number }[];
  bills: { id: string; name: string; amount_cents: number; due_day: number; category_id: string | null; is_active: boolean }[];
  members: { id: string; full_name: string }[];
  creditProfiles: { user_id: string; utilization_pct: number | null }[];
  creditHistory: { user_id: string; score: number; recorded_date: string; created_at_ms: number }[];
}

export type NotificationType =
  | "bill_due"
  | "budget_exceeded"
  | "goal_milestone"
  | "large_transaction"
  | "investment_opportunity"
  | "sync_error"
  | "credit_alert";

export interface NotificationDraft {
  /** Identifies the event; a notification is only ever created once per key. */
  dedupe_key: string;
  type: NotificationType;
  severity: "info" | "warning" | "success";
  title: string;
  message: string;
  related_entity_type: string;
  related_entity_id: string;
}

export interface RuleResults {
  notifications: NotificationDraft[];
  /** Goals whose highest announced milestone moved up. */
  goalMilestones: { goalId: string; pct: number }[];
}

const MILESTONES = [25, 50, 75, 100];

export function evaluateRules(data: HouseholdData, today: string): RuleResults {
  const notifications: NotificationDraft[] = [];
  const goalMilestones: { goalId: string; pct: number }[] = [];
  const accountIds = new Set(data.accounts.map((a) => a.id));
  const transactions = data.transactions.filter((t) => accountIds.has(t.account_id));
  const debits = transactions.filter((t) => t.direction === "debit");
  const thisMonth = monthStart(today);
  const nextMonth = addMonths(thisMonth, 1);
  const period = today.slice(0, 7);

  // Budgets: a heads-up at 90% of the limit and another once it's exceeded, once each per month.
  const categoryNames = new Map(data.categories.map((c) => [c.id, c.name]));
  for (const budget of data.budgets.filter((b) => b.month === thisMonth)) {
    const limit = budget.amount_limit_cents;
    if (limit <= 0) continue;
    const spent = debits
      .filter((t) => t.category_id === budget.category_id && t.date >= thisMonth && t.date < nextMonth)
      .reduce((sum, t) => sum + t.amount_cents, 0);
    const pct = spent / limit;
    const category = categoryNames.get(budget.category_id) ?? "this category";
    const related = { related_entity_type: "budget", related_entity_id: budget.id };
    if (pct >= 1) {
      notifications.push({
        dedupe_key: `budget:${budget.id}:${period}:exceeded`,
        type: "budget_exceeded",
        severity: "warning",
        title: `Budget exceeded: ${category}`,
        message: `You've spent ${money(spent)} of your ${money(limit)} ${category} budget this month (${wholePercent(pct * 100)}%).`,
        ...related,
      });
    } else if (pct >= BUDGET_WARNING_PCT) {
      notifications.push({
        dedupe_key: `budget:${budget.id}:${period}:warning`,
        type: "budget_exceeded",
        severity: "info",
        title: `Approaching budget limit: ${category}`,
        message: `You've used ${wholePercent(pct * 100)}% of your ${money(limit)} ${category} budget this month (${money(spent)} spent).`,
        ...related,
      });
    }
  }

  // Goals: announce only the highest newly reached milestone.
  for (const goal of data.goals) {
    if (goal.target_amount_cents <= 0) continue;
    const pct = (goal.current_amount_cents / goal.target_amount_cents) * 100;
    const reached = MILESTONES.filter((m) => pct >= m && m > goal.last_milestone_pct);
    if (reached.length === 0) continue;
    const milestone = Math.max(...reached);
    notifications.push({
      dedupe_key: `goal:${goal.id}:${milestone}`,
      type: "goal_milestone",
      severity: milestone === 100 ? "success" : "info",
      title: milestone === 100 ? `Goal complete: ${goal.name}!` : `${milestone}% of the way to ${goal.name}`,
      message: `${goal.name} is at ${money(goal.current_amount_cents)} of ${money(goal.target_amount_cents)} (${wholePercent(pct)}%).`,
      related_entity_type: "goal",
      related_entity_id: goal.id,
    });
    goalMilestones.push({ goalId: goal.id, pct: milestone });
  }

  // Bills due within the lead time, unless a matching payment has already posted.
  for (const bill of data.bills.filter((b) => b.is_active)) {
    let due = withDay(today, bill.due_day);
    if (due < today) due = addMonths(due, 1);
    const daysUntil = daysBetween(today, due);
    if (daysUntil < 0 || daysUntil > UPCOMING_BILL_LEAD_DAYS) continue;

    const periodStart = addMonths(due, -1);
    const alreadyPaid = debits.some(
      (t) =>
        t.date >= periodStart &&
        t.date <= addDays(due, 1) &&
        t.amount_cents >= bill.amount_cents * 0.85 &&
        t.amount_cents <= bill.amount_cents * 1.15 &&
        (!bill.category_id || t.category_id === bill.category_id),
    );
    if (alreadyPaid) continue;

    const when = daysUntil === 0 ? "today" : daysUntil === 1 ? "tomorrow" : `in ${daysUntil} days`;
    notifications.push({
      dedupe_key: `bill:${bill.id}:${due}`,
      type: "bill_due",
      severity: daysUntil <= 1 ? "warning" : "info",
      title: `${bill.name} due ${when}`,
      message: `${money(bill.amount_cents)} for ${bill.name} is due on ${shortDate(due)}.`,
      related_entity_type: "recurring_bill",
      related_entity_id: bill.id,
    });
  }

  // Large debits in the last two days.
  const since = addDays(today, -2);
  for (const t of debits.filter((t) => t.amount_cents >= LARGE_TRANSACTION_THRESHOLD_CENTS && t.date >= since)) {
    notifications.push({
      dedupe_key: `large_txn:${t.id}`,
      type: "large_transaction",
      severity: "info",
      title: `Large transaction: ${money(t.amount_cents)}`,
      message: `${t.name} on ${shortDate(t.date)} was ${money(t.amount_cents)}.`,
      related_entity_type: "transaction",
      related_entity_id: t.id,
    });
  }

  // Savings balances above the idle-cash threshold, at most once a month per account.
  for (const account of data.accounts) {
    if (account.type !== "savings" || account.current_balance_cents < IDLE_CASH_INVESTMENT_THRESHOLD_CENTS) continue;
    notifications.push({
      dedupe_key: `idle_cash:${account.id}:${period}`,
      type: "investment_opportunity",
      severity: "info",
      title: `Idle cash in ${account.name}`,
      message:
        `${account.name} has ${money(account.current_balance_cents)}, above your ` +
        `${money(IDLE_CASH_INVESTMENT_THRESHOLD_CENTS)} idle-cash threshold. Once your emergency fund is covered, ` +
        "consider moving the excess into an investment or high-yield account.",
      related_entity_type: "account",
      related_entity_id: account.id,
    });
  }

  // Credit: high utilization (monthly) and a drop of 15+ points between the last two scores.
  const names = new Map(data.members.map((m) => [m.id, m.full_name]));
  for (const profile of data.creditProfiles) {
    const name = names.get(profile.user_id) ?? "a household member";
    const related = { related_entity_type: "credit_profile", related_entity_id: profile.user_id };
    if (profile.utilization_pct != null && profile.utilization_pct > 30) {
      notifications.push({
        dedupe_key: `credit_util:${profile.user_id}:${period}`,
        type: "credit_alert",
        severity: "warning",
        title: `High credit utilization for ${name}`,
        message:
          `${name}'s reported utilization is ${wholePercent(profile.utilization_pct)}%, ` +
          "above the 30% level that most scoring models penalize. Paying down balances " +
          "before the statement closes date can help.",
        ...related,
      });
    }

    const [latest, previous] = data.creditHistory
      .filter((h) => h.user_id === profile.user_id)
      .sort((a, b) => b.recorded_date.localeCompare(a.recorded_date) || b.created_at_ms - a.created_at_ms);
    if (latest && previous && previous.score - latest.score >= 15) {
      notifications.push({
        dedupe_key: `credit_drop:${profile.user_id}:${latest.recorded_date}`,
        type: "credit_alert",
        severity: "warning",
        title: `Credit score drop for ${name}`,
        message:
          `${name}'s score fell from ${previous.score} to ${latest.score}. ` +
          "Check for new derogatory marks, hard inquiries, or a spike in utilization.",
        ...related,
      });
    }
  }

  return { notifications, goalMilestones };
}
